### ISSUE TITLE
false positive in depot_save_stack

### ISSUE THREAD

**Issue by dvyukov** [google/kasan#37]
Here are two crashes from syzkaller:
https://gist.githubusercontent.com/dvyukov/56c471c2cbf4aaa97aceb415bbb4e874/raw/40823c87465a70d03e05e668cfa82e1e1d16632b/gistfile1.txt
https://gist.githubusercontent.com/dvyukov/d821a21df08e71c1acb396d9ad951752/raw/b699c035ab1f54d6b9d495b50ae13ef8144990c0/gistfile1.txt

Both have similar false positive reports:

```
[  322.479239] ==================================================================
[  322.480040] BUG: KASAN: stack-out-of-bounds in memcmp+0xe9/0x150 at addr ffff880038bb74e8
[  322.480040] Read of size 1 by task syz-executor/553
[  322.480040] page:ffffea0000e2edc0 count:0 mapcount:0 mapping:          (null) index:0x0
[  322.487260] flags: 0xfffe0000000000()
[  322.487260] page dumped because: kasan: bad access detected
[  322.487260] CPU: 2 PID: 553 Comm: syz-executor Not tainted 4.7.0-rc5+ #28
[  322.487260] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS Bochs 01/01/2011
[  322.487260]  ffffffff880b58e0 ffff880038bb7310 ffffffff82cc01af ffffffff38bb73a0
[  322.487260]  fffffbfff1016b1c ffff880038bb73a0 ffff880038bb74e8 ffff880038bb7548
[  322.487260]  ffff880038bb7540 00000000000000b0 ffff880038bb7390 ffffffff817b9a03
[  322.487260] Call Trace:
[  322.487260]  [<ffffffff82cc01af>] dump_stack+0x12e/0x18f
[  322.487260]  [<ffffffff817b9a03>] kasan_report_error+0x573/0x5a0
[  322.487260]  [<ffffffff8151e6e3>] ? __module_text_address+0x13/0x150
[  322.487260]  [<ffffffff817b9a6e>] __asan_report_load1_noabort+0x3e/0x40
[  322.487260]  [<ffffffff82cddcb9>] ? memcmp+0xe9/0x150
[  322.487260]  [<ffffffff82cddcb9>] memcmp+0xe9/0x150
[  322.487260]  [<ffffffff82d8d7e6>] depot_save_stack+0x176/0x5b0
[  322.487260]  [<ffffffff858c0e24>] ? skb_free_head+0x74/0xb0
[  322.487260]  [<ffffffff817b8951>] save_stack+0xb1/0xd0
...
[  322.554627] Memory state around the buggy address:
[  322.554627]  ffff880038bb7380: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[  322.554627]  ffff880038bb7400: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[  322.554627] >ffff880038bb7480: 00 00 00 00 00 00 00 00 00 00 00 00 00 f4 f4 00
[  322.554627]                                                           ^
[  322.554627]  ffff880038bb7500: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[  322.554627]  ffff880038bb7580: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

f4 is leftover of stack frames.

**Comment by dvyukov** [comment]
At some point I saw lots of false positives in sctp interrupt handlers, which turned out to be related to jprobes. I fixed it with:

```
diff --git a/arch/x86/kernel/kprobes/core.c b/arch/x86/kernel/kprobes/core.c
index 7847e5c..9198c18 100644
--- a/arch/x86/kernel/kprobes/core.c
+++ b/arch/x86/kernel/kprobes/core.c
@@ -1080,6 +1080,8 @@ void jprobe_return(void)
 {
        struct kprobe_ctlblk *kcb = get_kprobe_ctlblk();
 
+       kasan_unpoison_shadow((char*)&kcb - 32, (char*)kcb->jprobe_saved_sp - (char*)&kcb + 64);
+
        asm volatile (
 #ifdef CONFIG_X86_64
                        "       xchg   %%rbx,%%rsp      \n"
```

Later the false positives disappeared, so I dropped the change.
Probably we still need it something along these lines. Not sure about other stores to `%rsp` in `arch/x86/kernel/kprobes/{core,opt}.c`.

**Comment by xairy** [comment]
This looks fixed by 9f7d416c36124667c406978bcb39746589c35d7f.
